chia/full_node/mempool.py

```diff
@@ -90,66 +172,76 @@ def remove_from_pool(self, items: List[bytes32], reason: MempoolRemoveReason) ->
+                        name = bytes32(row[0])
+                        internal_item = self._items[name]
+                        item = MempoolItemInfo(int(row[1]), int(row[2]), internal_item.height_added_to_mempool)
+                        removed_items.append(item)
+
+        for name in items:
+            self._items.pop(name)
+
+        for spend_bundle_ids in chunks(items, SQLITE_MAX_VARIABLE_NUMBER):
+            args = ",".join(["?"] * len(spend_bundle_ids))
+            with self._db_conn:
+                self._db_conn.execute(f"DELETE FROM tx WHERE name in ({args})", spend_bundle_ids)
+                self._db_conn.execute(f"DELETE FROM spends WHERE tx in ({args})", spend_bundle_ids)
+
+        if reason != MempoolRemoveReason.BLOCK_INCLUSION:
+            info = FeeMempoolInfo(
+                self.mempool_info, self.total_mempool_cost(), self.total_mempool_fees(), datetime.now()
+            )
+            for iteminfo in removed_items:
+                self.fee_estimator.remove_mempool_item(info, iteminfo)
 
     def add_to_pool(self, item: MempoolItem) -> None:
         """
         Adds an item to the mempool by kicking out transactions (if it doesn't fit), in order of increasing fee per cost
         """
 
+        assert item.fee < MEMPOOL_ITEM_FEE_LIMIT
         assert item.npc_result.conds is not None
 
-        while self.at_full_capacity(item.cost):
-            # Val is Dict[hash, MempoolItem]
-            fee_per_cost, val = self._sorted_spends.peekitem(index=0)
-            to_remove: MempoolItem = list(val.values())[0]
-            self.remove_from_pool([to_remove.name], MempoolRemoveReason.POOL_FULL)
-
-        self._spends[item.name] = item
-
-        # _sorted_spends is Dict[float, Dict[bytes32, MempoolItem]]
-        if item.fee_per_cost not in self._sorted_spends:
-            self._sorted_spends[item.fee_per_cost] = {}
+        # TODO: this block could be simplified by removing all items in a single
+        # SQL query. Or at least figure out which items to remove and then
+        # remove them all in a single call to remove_from_pool()
+        with self._db_conn:
+            while self.at_full_capacity(item.cost):
+                # pick the item with the lowest fee per cost
+                cursor = self._db_conn.execute("SELECT name FROM tx ORDER BY fee_per_cost ASC LIMIT 1")
+                name = bytes32(cursor.fetchone()[0])
+                self.remove_from_pool([name], MempoolRemoveReason.POOL_FULL)
+
+            self._db_conn.execute(
+                "INSERT INTO tx VALUES(?, ?, ?, ?)", (item.name, item.cost, item.fee, item.assert_height)
+            )
 
-        self._sorted_spends[item.fee_per_cost][item.name] = item
+            all_coin_spends = [(s.coin_id, item.name) for s in item.npc_result.conds.spends]
+            for spends in chunks(all_coin_spends, SQLITE_MAX_VARIABLE_NUMBER):
+                self._db_conn.executemany("INSERT INTO spends VALUES(?, ?)", all_coin_spends)
```

Comment:
typo

Reply:
yes, but I will remove the loop, so this typo will become correct